lib/portage/package/ebuild/fetch.py

```diff
@@ -201,6 +226,19 @@ def _userpriv_test_write_file(settings, file_path):
     if rval is not None:
         return rval
 
+    # Return early in order to avoid an unecessary spawn and event loop
+    # recursion when privileges do not need to be dropped.
+    if not (
+        "userfetch" in settings.features
+        and os.getuid() == 0
+        and portage_gid
+        and portage_uid
+        and hasattr(os, "setgroups")
+    ):
+        rval = os.access(os.path.dirname(file_path), os.W_OK)
+        _userpriv_test_write_file_cache[file_path] = rval
+        return rval
+
```

Comment:
I've triggered an `RuntimeError: This event loop is already running` error here by running `sudo ebuild <ebuild> manifest clean install`:
```
Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.12/ebuild", line 444, in <module>
    main()
  File "/usr/lib/python-exec/python3.12/ebuild", line 405, in main
    a = portage.doebuild(
        ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/doebuild.py", line 1387, in doebuild
    return not digestgen(mysettings=mysettings, myportdb=mydbapi)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/digestgen.py", line 161, in digestgen
    if not fetch({myfile: uris}, mysettings):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/fetch.py", line 835, in fetch
    return asyncio.run(
           ^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/util/futures/_asyncio/__init__.py", line 115, in run
    return _safe_loop().run_until_complete(coro)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/util/_eventloop/asyncio_event_loop.py", line 173, in _run_until_complete
    return self._loop.run_until_complete(future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/fetch.py", line 1199, in async_fetch
    _ensure_distdir(mysettings, mysettings["DISTDIR"])
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/fetch.py", line 281, in _ensure_distdir
    if _userpriv_test_write_file(settings, write_test_file):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/fetch.py", line 248, in _userpriv_test_write_file
    returncode = _spawn_fetch(settings, args)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/package/ebuild/fetch.py", line 173, in _spawn_fetch
    rval = spawn_func(args, env=env, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/process.py", line 730, in spawn
    retval = loop.run_until_complete(pid.wait())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/portage/util/_eventloop/asyncio_event_loop.py", line 173, in _run_until_complete
    return self._loop.run_until_complete(future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 663, in run_until_complete
    self._check_running()
  File "/usr/lib/python3.12/asyncio/base_events.py", line 622, in _check_running
    raise RuntimeError('This event loop is already running')
RuntimeError: This event loop is already running
```

Reply:
As an approximation, I've used `os.stat` to test for portage group write access.

This makes an assumption about the implementation details for DISTDIR write access with `FEATURES=userfetch`, since the approximation using `os.stat`  works with normal Unix permissions but ignores ACLs.

Reply:
We need to change the _drop_privs_userfetch function to an async coroutine function, since it calls _userpriv_test_write_file via _ensure_distdir (so all three functions need to be coroutine functions).

Reply:
Since the `_drop_privs_userfetch` function is always called before the event loop is running, I simply adapted it like this:
```diff
--- a/lib/portage/package/ebuild/fetch.py
+++ b/lib/portage/package/ebuild/fetch.py
@@ -114,3 +114,3 @@ def _drop_privs_userfetch(settings):
     try:
-        _ensure_distdir(settings, settings["DISTDIR"])
+        asyncio.run(_ensure_distdir(settings, settings["DISTDIR"]))
     except PortageException:
```